### PR TITLE
[improve][ci] Upgrade/Downgrade test

### PR DESCRIPTION
**Pull request by dlg99**
### Motivation

catch issues like https://github.com/apache/pulsar/issues/22051

### Modifications

Added a test for upgrade/downgrade

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change adds tests.

### Does this pull request potentially affect one of the following parts:

*NO*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/dlg99/pulsar/pull/18

### CODE REVIEW

**Comment by dlg99** [comment]
@lhotari ping

File: .github/workflows/pulsar-ci.yaml [hunk 1]
```diff
@@ -648,6 +651,18 @@ jobs:
...
+          docker pull apachepulsar/pulsar:2.10.6
+
+      - name: Pull apachepulsar/pulsar:3.0.5
+        run: |
+          docker pull apachepulsar/pulsar:3.0.5
+
+      - name: Pull alpine:3.20.1
+        run: |
+          docker pull alpine:3.20.1
+
```
Comment (lhotari) [inline .github/workflows/pulsar-ci.yaml:665]: This is unnecessary and additional overhead for most builds in this matrix. TestContainers will pull the images when it's necessary to do that.

**Review by lhotari — COMMENTED** [review]
Thank you @dlg99, this is very useful. I have provided review feedback to make this more generic and usable in maintenance branches.

**Review by lhotari — APPROVED** [review]
LGTM, great work @dlg99

File: tests/integration/src/test/java/org/apache/pulsar/tests/integration/upgrade/PulsarUpgradeDowngradeTest.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+                .withEnv("MAGIC_NUMBER", "42")
+                .withCreateContainerCmdModifier(createContainerCmd -> createContainerCmd
+                    .getHostConfig()
+                    .withBinds(Bind.parse("/pulsar/data:/pulsar/data")))
+                .withCommand("/bin/sh", "-c",
+                        "mkdir -p /pulsar/data && "
+                                + "chmod -R ug+rwx /pulsar/data && "
+                                + "chown -R 10000:0 /pulsar/data && "
+                                + "rm -rf /pulsar/data/* && "
+                                + "while true; do echo \"$MAGIC_NUMBER\" | nc -l -p 80; done");
```
Comment (lhotari) [inline tests/integration/src/test/java/org/apache/pulsar/tests/integration/upgrade/PulsarUpgradeDowngradeTest.java:75]: what is the purpose of this command?

**Pull request merged**